### PR TITLE
[SYCL] Split device images based on accuracy level provided in option

### PR DESCRIPTION
**Pull request by againull**
This PR reuses optional kernel features mechanism to split device image
based on accuracy level provided using -ffp-accuracy compilation option (introduced in PR#8280):
1. When frontend emits fp intrinsic call and attaches the maximum error
   attribute we also attach "sycl_used_aspects" metadata to the call
   instruction with a value which corresponds to high, medium, low, sycl
   or cuda. Mapping for those values is needed to be visible for SYCL
   device compiler only and we intentionally don't put those values to
   aspects enum because we don't need these aspects to be visible to 
   the user in this case  (because of the reasons described in Details below).
2. Make SYCLPropagateAspectsUsage to propagate sycl_used_aspects
   metadata from instructions to kernel.
3. Don't add internal aspects into the device requirements, because we
    don't need processing of these internal aspects (with negative values) in the SYCL RT.

Splitting functionality based on sycl_used_aspects metadata is available for free.

Details:
Currently accruracy level can be controlled using the following options.
For entire translation unit:
-ffp-accuracy=high
-ffp-accuracy=medium
-ffp-accuracy=low
-ffp-accuracy=sycl
-ffp-accuracy=cuda

For particular funcions in the translation unit:
-ffp-accuracy=low:sin,cos

Whenever frontend sees a math function in a kernel or a device function
it emits fp intrinsic call with attached callsite attribute indicating
value of the maximum error. llvm-spirv is going to translate this
builtins to regular __ocl intrinsics and translate callsite attribute to
decorator (which is a new spirv extension). If that extension is not supported
by the backend, it is going to emit an error. Error is emitted also in
the case if backend supports the extension but can't compile the kernel because
it doesn't have corresponding implemenation of math function complying with
required maximum error.

Aspects corrsponding to different levels of accuracy are not suitable in
this case because aforementioned options are sycl program compilation options, i.e.
it doesn't make sense to provide an opportunity to the user to write
something like this:
if (dev.has(aspect::ext_oneapi_fp_intrinsic_accuracy_high)) {
  /* submit kernel using high accuracy intrinsics */
}

But on our side we still would like to put kernels and device functions
to different images based on required accuracy level. It is necessary because
some backends may support, for example, low and medium accuracy but don't
support high accuracy.

### CODE REVIEW

**Review by elizabethandrews — COMMENTED** [review]
Please add a frontend test

**Comment by asudarsa** [comment]
Hi @againull 

Thanks for the PR.
I had some high level questions about this implementation:
1. This implementation seems to assume that the backend will either have support (or no support) for all functions with a specific level of accuracy. Is it possible that the backend may support 'high accuracy' implementations of some functions and 'low accuracy' implementations of some other functions? I am not sure if that scenario is supported here.
2. You have used '-1' to '-5' as metadata here. How can we guarantee that no other pass will insert metadata that clashes with this? i.e. How do we intend to document this?
3. Is it possible to use the existing attribute, instead of adding new aspect here? We might need to propagate attributes from call site to caller function. We do have a mechanism to easily split device code based on attributes. 

Thanks again

File: clang/lib/CodeGen/CGCall.cpp [hunk 1]
```diff
@@ -5620,8 +5637,9 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
   // Emit the actual call/invoke instruction.
   llvm::CallBase *CI;
   if (!InvokeDest) {
-    if (!getLangOpts().FPAccuracyFuncMap.empty() ||
-        !getLangOpts().FPAccuracyVal.empty()) {
+    if ((!getLangOpts().FPAccuracyFuncMap.empty() ||
+         !getLangOpts().FPAccuracyVal.empty()) &&
+        isa_and_nonnull<FunctionDecl>(TargetDecl)) {
```
Comment (asudarsa) [inline clang/lib/CodeGen/CGCall.cpp:5642]: What is the relevance of this change w.r.t the overall scope of this PR? Sorry if I am missing something here.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 2]
```diff
@@ -513,12 +514,17 @@ static CallInst *CreateBuiltinCallWithAttr(CodeGenFunction &CGF, StringRef Name,
   // TODO: Replace AttrList with a single attribute. The call can only have a
   // single FPAccuracy attribute.
   llvm::AttributeList AttrList;
+  // "sycl_used_aspects" metadata associated with the call.
+  SmallVector<llvm::Metadata *, 4> AspectsMD;
```
Comment (asudarsa) [inline clang/lib/CodeGen/CGBuiltin.cpp:518]: Does this need to be a vector? I am not sure if we can multiple such MD associated with a single call. Please note the TODOassociated with the AttrList above. 

Thanks

File: clang/lib/CodeGen/CGCall.cpp [hunk 3]
```diff
@@ -1846,8 +1847,18 @@ static llvm::fp::FPAccuracy convertFPAccuracy(StringRef FPAccuracyStr) {
       .Case("cuda", llvm::fp::FPAccuracy::CUDA);
 }
 
+static int32_t convertFPAccuracyToAspect(StringRef FPAccuracyStr) {
```
Comment (asudarsa) [inline clang/lib/CodeGen/CGCall.cpp:1850]: Do we need to add an assert here to ensure this function is called with appropriate FPAccuracyStr?

Thanks

File: clang/lib/CodeGen/CGCall.cpp [hunk 4]
```diff
@@ -1864,6 +1875,9 @@ void CodeGenModule::getDefaultFunctionFPAccuracyAttributes(
           ID, FuncType, convertFPAccuracy(FuncMapIt->second));
       assert(!FPAccuracyVal.empty() && "A valid accuracy value is expected");
       FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+      if (getLangOpts().SYCLIsDevice)
```
Comment (asudarsa) [inline clang/lib/CodeGen/CGCall.cpp:1878]: Do we need this check here?

Thanks

File: sycl/test/optional_kernel_features/fp-accuracy.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
+// RUN: %clangxx %s -o %test.bc -ffp-accuracy=high:sin,sqrt -ffp-accuracy=medium:cos -ffp-accuracy=low:tan -ffp-accuracy=cuda:exp,acos -ffp-accuracy=sycl:log,asin  -fno-math-errno  -fsycl -fsycl-device-only
```
Comment (asudarsa) [inline sycl/test/optional_kernel_features/fp-accuracy.cpp:1]: Is it possible to test if the aspects got correctly propagated to the calling functions?

Thanks

File: clang/test/CodeGenSYCL/fp-accuracy.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      (void)sqrt(Value);
+    });
+  });
+  return 0;
+}
+
+// CHECK: [[ASPECT1]] = !{i32 -1}
+// CHECK: [[ASPECT2]] = !{i32 -2}
+// CHECK: [[ASPECT3]] = !{i32 -3}
+// CHECK: [[ASPECT4]] = !{i32 -5}
```
Comment (asudarsa) [inline clang/test/CodeGenSYCL/fp-accuracy.cpp:101]: nit: can be rename these to match aspect name and the value?

**Review by asudarsa — CHANGES_REQUESTED** [review]
I have added few high-level questions and some code comments. please address.

Thanks

File: clang/lib/CodeGen/CGCall.cpp (hunk 3 above)
Comment (againull) [reply clang/lib/CodeGen/CGCall.cpp:1850]: Added assert, thanks.

**Comment by againull** [comment]
> Please add a frontend test

Added frontend test, thanks!

**Comment by againull** [comment]
We mostly discussed this offline with @asudarsa , will duplicate this info here.

> 1. This implementation seems to assume that the backend will either have support (or no support) for all functions with a specific level of accuracy. Is it possible that the backend may support 'high accuracy' implementations of some functions and 'low accuracy' implementations of some other functions? I am not sure if that scenario is supported here.

We have agreed about this level of granularity after discussion with Gregory Lueck. Yes, it possible that  we may support 'high accuracy' implementations of some functions and 'low accuracy' implementations of some other functions, but splitting image per each function or other greater granularity doesn't seem feasible. Backends will have to report an error if required accuracy for particular math function is not supported.

> 2. You have used '-1' to '-5' as metadata here. How can we guarantee that no other pass will insert metadata that clashes with this? i.e. How do we intend to document this?

Since this internal aspect values are not exposed to the user and only part of implementation details then it seems reasonable to describe this in https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md
I have added description there.

> 3. Is it possible to use the existing attribute, instead of adding new aspect here? We might need to propagate attributes from call site to caller function. We do have a mechanism to easily split device code based on attributes.
> 
> Thanks again

Unfortunately, I am not sure that we can do that. Attribute contains only the value of error which may be different for each math function (for example for sycl or cuda accuracy), so basically attribute doesn't contain information about level of accuracy anymore, it only contains error value.

**Comment by againull** [comment]
@sergey-semenov Could you please take a look to the PR as well.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 7]
```diff
@@ -16,6 +16,7 @@
 #include "CGObjCRuntime.h"
 #include "CGOpenCLRuntime.h"
 #include "CGRecordLayout.h"
+#include "CGSYCLRuntime.h"
```
Comment (elizabethandrews) [inline clang/lib/CodeGen/CGBuiltin.cpp:19]: Do you need this include here? I don't see you using internal aspects in this file.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 8]
```diff
@@ -22144,7 +22150,8 @@ llvm::CallInst *CodeGenFunction::EmitFPBuiltinIndirectCall(
     // Even if the current function doesn't have a clang builtin, create
     // an 'fpbuiltin-max-error' attribute for it; unless it's marked with
     // an NoBuiltin attribute.
-    if (!FD->hasAttr<NoBuiltinAttr>()) {
+    if (!FD->hasAttr<NoBuiltinAttr>() &&
+        FD->getNameInfo().getName().isIdentifier()) {
```
[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/lib/CodeGen/CGBuiltin.cpp:22155]: If it's a CXXContructorDecl then it's not an identifier.

File: clang/lib/CodeGen/CGCall.cpp (hunk 1 above)
Comment (zahiraam) [reply clang/lib/CodeGen/CGCall.cpp:5643]: As we discussed offline for the 2 tests that you were looking at, I didn't see the need for this fix. Not quite sure what this is doing?

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 9]
```diff
@@ -515,16 +515,17 @@ static CallInst *CreateBuiltinCallWithAttr(CodeGenFunction &CGF, StringRef Name,
   // single FPAccuracy attribute.
   llvm::AttributeList AttrList;
   // "sycl_used_aspects" metadata associated with the call.
-  SmallVector<llvm::Metadata *, 4> AspectsMD;
+  llvm::Metadata *AspectMD;
```
Comment (asudarsa) [inline clang/lib/CodeGen/CGBuiltin.cpp:518]: You might run into static analyzer issues due to not initializing this.

**Review by asudarsa — APPROVED** [review]
Changes look good. Thanks

File: clang/lib/CodeGen/CGCall.cpp (hunk 1 above)
Comment (againull) [reply clang/lib/CodeGen/CGCall.cpp]: Oh, I've rechecked the tests and indeed this change is redundant. Removed this change from PR. Thanks.

File: clang/lib/CodeGen/CGCall.cpp [hunk 10]
```diff
@@ -5620,8 +5638,8 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
   // Emit the actual call/invoke instruction.
   llvm::CallBase *CI;
   if (!InvokeDest) {
-    if (!getLangOpts().FPAccuracyFuncMap.empty() ||
-        !getLangOpts().FPAccuracyVal.empty()) {
+    if ((!getLangOpts().FPAccuracyFuncMap.empty() ||
+         !getLangOpts().FPAccuracyVal.empty())) {
```
Comment (zahiraam) [inline clang/lib/CodeGen/CGCall.cpp:5642]: Why double parentheses?

File: clang/test/CodeGenSYCL/fp-accuracy.cpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
+// RUN: %clang_cc1  -fsycl-is-device -ffp-builtin-accuracy=high:sin,sqrt -ffp-builtin-accuracy=medium:cos -ffp-builtin-accuracy=low:tan -ffp-builtin-accuracy=cuda:exp,acos -ffp-builtin-accuracy=sycl:log,asin -emit-llvm -triple spir64-unknown-unknown -disable-llvm-passes %s -o - | FileCheck %s
```
Comment (zahiraam) [inline clang/test/CodeGenSYCL/fp-accuracy.cpp:1]: Can you add a run line where a TU value of accuracy is used and a another one with a mix of TU accuracy and function specific ones?

File: sycl/test/optional_kernel_features/fp-accuracy.cpp (hunk 5 above)
Comment (zahiraam) [inline sycl/test/optional_kernel_features/fp-accuracy.cpp:1]: Same request here for additional RUN lines.

File: clang/lib/CodeGen/CGCall.cpp (hunk 10 above)
Comment (againull) [reply clang/lib/CodeGen/CGCall.cpp]: Sorry, fixed.

File: clang/test/CodeGenSYCL/fp-accuracy.cpp (hunk 11 above)
Comment (againull) [reply clang/test/CodeGenSYCL/fp-accuracy.cpp]: Added additional run lines according to your suggestion.

File: sycl/test/optional_kernel_features/fp-accuracy.cpp (hunk 5 above)
Comment (againull) [reply sycl/test/optional_kernel_features/fp-accuracy.cpp]: Added additional run lines according to your suggestion.

**Review by zahiraam — APPROVED** [review]
LGTM. Thanks.

**Comment by AlexeySachkov** [comment]
> We mostly discussed this offline with @asudarsa , will duplicate this info here.
> 
> > 3. Is it possible to use the existing attribute, instead of adding new aspect here? We might need to propagate attributes from call site to caller function. We do have a mechanism to easily split device code based on attributes.
> > 
> > Thanks again
> 
> Unfortunately, I am not sure that we can do that. Attribute contains only the value of error which may be different for each math function (for example for sycl or cuda accuracy), so basically attribute doesn't contain information about level of accuracy anymore, it only contains error value.

I'm confused here: why it is possible to propagate `-1` in a metadata from intrinsic to a kernel, but isn't possible to propagate `"uses-accuracy=low"` (or something like that) ?

Suggested approach does looks like a hack, a bit and I think that we should have some generic infrastructure for propagating attributes/metadata, which are not aspects. There are a few examples already where it is needed: in `sycl-post-link` we are trying to understand which kernels are using `assert`; accuracy levels from this PR; there are plans to do device code split based on data types from joint matrix extension and they won't fit into aspects as well, due to complexity.

I don't want to block this PR, but I would like to ensure that there is a path to refactor accuracy levels handling from aspects to generic attributes/metadata propagation.

**Comment by againull** [comment]
Hello @Fznamznon, could you please help to review this PR. Originally @elizabethandrews was looking at this from [intel/dpcpp-cfe-reviewers](https://github.com/orgs/intel/teams/dpcpp-cfe-reviewers) group. But unfortunately she is on vacation and other members are on vacation as well. Formally @zahiraam is not  dpcpp-cfe-reviewers group, so I am not sure if I  can treat her approval as green signal from CFE group.

**Comment by againull** [comment]
HIP AMDGPU failure is an infrastructure issue which is unrelated to this PR.

File: clang/lib/CodeGen/CGSYCLRuntime.h [hunk 12]
```diff
@@ -23,6 +23,19 @@ namespace CodeGen {
 
 class CodeGenModule;
 
+// These aspects are internal and used for device image splitting purposes only.
+// They are not exposed to the DPCPP users through "aspect" enum. That's why
```
Comment (Fznamznon) [inline clang/lib/CodeGen/CGSYCLRuntime.h:27]: We're still ~dreaming~ thinking about upstream, let's make it a teeny tiny bit easier.
```suggestion
// They are not exposed to the SYCL users through "aspect" enum. That's why
```

File: clang/lib/CodeGen/CGSYCLRuntime.h [hunk 13]
```diff
@@ -23,6 +23,19 @@ namespace CodeGen {
...
+// they are intentionally assigned negative values to filter them out at the
+// stage of embedding used aspects as device requirements to the executable.
+// We don't pass these internal aspects to the DPCPP RT.
```
Comment (Fznamznon) [inline clang/lib/CodeGen/CGSYCLRuntime.h:30]:
```suggestion
// We don't pass these internal aspects to the SYCL RT.
```

File: clang/test/CodeGenSYCL/fp-accuracy.cpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
+// RUN: %clang_cc1  -fsycl-is-device -ffp-builtin-accuracy=high:sin,sqrt -ffp-builtin-accuracy=medium:cos -ffp-builtin-accuracy=low:tan -ffp-builtin-accuracy=cuda:exp,acos -ffp-builtin-accuracy=sycl:log,asin -emit-llvm -triple spir64-unknown-unknown -disable-llvm-passes %s -o - | FileCheck --check-prefix CHECK-FUNC %s
+// RUN: %clang_cc1  -fsycl-is-device -ffp-builtin-accuracy=high -emit-llvm -triple spir64-unknown-unknown -disable-llvm-passes %s -o - | FileCheck --check-prefix CHECK-TU %s
+// RUN: %clang_cc1  -fsycl-is-device -ffp-builtin-accuracy=medium -ffp-builtin-accuracy=high:sin,sqrt -ffp-builtin-accuracy=medium:cos -ffp-builtin-accuracy=cuda:exp -ffp-builtin-accuracy=sycl:log -emit-llvm -triple spir64-unknown-unknown -disable-llvm-passes %s -o - | FileCheck --check-prefix CHECK-MIX %s
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/fp-accuracy.cpp:3]: Default opt level for SYCL is O0 as well as for C++. Please apply to all run lines.
```suggestion
// RUN: %clang_cc1  -fsycl-is-device -ffp-builtin-accuracy=medium -ffp-builtin-accuracy=high:sin,sqrt -ffp-builtin-accuracy=medium:cos -ffp-builtin-accuracy=cuda:exp -ffp-builtin-accuracy=sycl:log -emit-llvm -triple spir64-unknown-unknown %s -o - | FileCheck --check-prefix CHECK-MIX %s
```

File: clang/test/CodeGenSYCL/fp-accuracy.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+// Tests that sycl_used_aspects metadata is attached to the fpbuiltin call based on -ffp-accuracy option.
+
+#include "Inputs/sycl.hpp"
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/fp-accuracy.cpp:7]: Please add tests according to https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ContributeToDPCPP.md#dpc-clang-fe-tests .

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/lib/CodeGen/CGBuiltin.cpp:22154]: Only functions are marked with this attribute. So, we are making sure that FD is not CXXContructorDecl.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 16]
```diff
@@ -22155,7 +22163,11 @@ llvm::CallInst *CodeGenFunction::EmitFPBuiltinIndirectCall(
               .Case("frem", llvm::Intrinsic::fpbuiltin_frem)
               .Case("sincos", llvm::Intrinsic::fpbuiltin_sincos)
               .Case("exp10", llvm::Intrinsic::fpbuiltin_exp10)
-              .Case("rsqrt", llvm::Intrinsic::fpbuiltin_rsqrt);
+              .Case("rsqrt", llvm::Intrinsic::fpbuiltin_rsqrt)
+              .Default(0);
+      if (!FPAccuracyIntrinsicID) {
```
[the remark this answers is not in the document]
Comment (againull) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Changed the code to avoid else-after-return.

File: clang/lib/CodeGen/CGSYCLRuntime.h (hunk 12 above)
Comment (againull) [reply clang/lib/CodeGen/CGSYCLRuntime.h]: Fixed.

File: clang/lib/CodeGen/CGSYCLRuntime.h (hunk 13 above)
Comment (againull) [reply clang/lib/CodeGen/CGSYCLRuntime.h]: Fixed.

File: clang/test/CodeGenSYCL/fp-accuracy.cpp (hunk 14 above)
Comment (againull) [reply clang/test/CodeGenSYCL/fp-accuracy.cpp]: Applied.

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/CodeGen/CGBuiltin.cpp]: It is a bit unclear to me how constructor may end up in a function named `EmitFPBuiltinIndirectCall`. I suppose `FD` represents the outer function, right?

[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Correct. FD represents the outer context/function.

**Review by Fznamznon — APPROVED** [review]

**Pull request merged**